src/libaktualizr/utilities/utils.cc

```diff
@@ -589,6 +588,113 @@ void Utils::writeArchive(const std::map<std::string, std::string> &entries, std:
   }
 }
 
+/* Removing a file from an archive isn't possible in the obvious sense. The only
+ * way to do so in practice is to create a new archive, copy everything you
+ * _don't_ want to remove, and then replace the old archive with the new one.
+ */
+void Utils::removeFileFromArchive(const boost::filesystem::path &archive_path, const std::string &filename) {
+  std::ifstream as_in(archive_path.c_str(), std::ios::in | std::ios::binary);
+  if (as_in.fail()) {
+    LOG_ERROR << "Unable to open provided provisioning archive " << archive_path << ": " << std::strerror(errno);
+    throw std::runtime_error("Unable to parse provisioning credentials");
+  }
+  TemporaryFile outfile;
+  std::ofstream as_out(outfile.PathString().c_str(), std::ios::out | std::ios::binary);
+  if (as_out.fail()) {
+    LOG_ERROR << "Unable to create file " << outfile.Path() << ": " << std::strerror(errno);
+    throw std::runtime_error("Unable to parse provisioning credentials");
+  }
+
+  struct archive *a_in = archive_read_new();
+  if (a_in == nullptr) {
+    LOG_ERROR << "archive error: could not initialize archive object";
+    throw std::runtime_error("archive error");
+  }
+  archive_read_support_filter_all(a_in);
+  archive_read_support_format_all(a_in);
+  auto state = std_::make_unique<archive_state>(std::ref(as_in));
+  int r = archive_read_open(a_in, reinterpret_cast<void *>(state.get()), nullptr, read_cb, nullptr);
+  if (r != ARCHIVE_OK) {
+    LOG_ERROR << "archive error: " << archive_error_string(a_in);
+    archive_read_free(a_in);
+    throw std::runtime_error("archive error");
+  }
+
+  struct archive *a_out = archive_write_new();
+  if (a_out == nullptr) {
+    LOG_ERROR << "archive error: could not initialize archive object";
```

Comment:
Don't we need `archive_read_free` here?

Reply:
Yep, good call. I did some RAII cleanup to this code and it should be better now. Also rebased.

Reply:
Cool! I always forget that we have StructGuard for such cases.